### PR TITLE
RavenDB-20029 Support CGroup V2

### PR DESCRIPTION
**Pull request by haludi**
### Issue link
https://issues.hibernatingrhinos.com/issue/RavenDB-20029

### Additional description
Checking which version of CGroup the system uses and use its paths

### Type of change
- Bug fix

### How risky is the change?
- Moderate 

### Backward compatibility
- Non breaking change

### Is it platform specific issue?
- Yes. Linux only.

### Documentation update
- No documentation update is needed 

### Testing by Contributor
- It has been verified by manual testing

### Testing by RavenDB QA team
- No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?
- No

### UI work
- No UI work is needed

### CODE REVIEW

File: src/Sparrow.Server/Platform/Posix/CGroupHelper.cs [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+            {
+                path = FindCGroupPathForMemory();
+            }
+            catch (Exception e)
+            {
+                if (Logger.IsOperationsEnabled)
+                    Logger.Operations("Failed to get CGroup path for memory", e);
+            }
+
+            return new CachedPath(path, DateTime.UtcNow);
```
Comment (ayende) [inline src/Sparrow.Server/Platform/Posix/CGroupHelper.cs:116]: Is there is an error, that will return `null`, is the rest of the code ready to handle that?

File: src/Sparrow.Server/Platform/Posix/CGroupHelper.cs [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+            //If the cgroup changed and the old cgroup was removed
+            basePath.Value.Deprecate();
+            return ReadValue(getBasePath, file, checkSpecialValues, false);
+        }
+    }
+
+    protected abstract string FindCGroupPathForMemory();
+    
+    //51 34 0:46 / /sys/fs/cgroup/hugetlb rw,nosuid,nodev,noexec,relatime shared:27 - cgroup cgroup rw,hugetlb
+    private static readonly Regex FindHierarchyMountReg = new Regex(@"^(?:\S+\s+){3}(?<mountroot>\S+)\s+(?<mountpath>\S+).* - (?<filesystemType>\S+)\s+\S+\s+(?:(?<options>[^,]+),?)+$");
```
Comment (ayende) [inline src/Sparrow.Server/Platform/Posix/CGroupHelper.cs]: This should probably be compiled

File: src/Sparrow.Server/Platform/Posix/CGroupHelper.cs (hunk 1 above)
Comment (haludi) [reply src/Sparrow.Server/Platform/Posix/CGroupHelper.cs:116]: Yes, the rest of the code is checking that
https://github.com/ravendb/ravendb/pull/16339/files#diff-f9e036622d8a33cd68e047f4902e971b72edb349b56cbbe0e42d4395fb43a18eR99

File: src/Sparrow.Server/Platform/Posix/CGroupHelper.cs (hunk 2 above)
Comment (haludi) [reply src/Sparrow.Server/Platform/Posix/CGroupHelper.cs]: Sorry, I am not sure what you mean

Comment (ayende) [reply src/Sparrow.Server/Platform/Posix/CGroupHelper.cs]: https://learn.microsoft.com/en-us/dotnet/standard/base-types/compilation-and-reuse-in-regular-expressions

Comment (haludi) [reply src/Sparrow.Server/Platform/Posix/CGroupHelper.cs]: That is very nice.
I added the flag

**Review by grisha-kotler — APPROVED** [review]

**Comment by gregolsky** [comment]
Can we add a smoke test (happy path) that would run on Linux only and try to read those values?

File: src/Sparrow.Server/Platform/Posix/CGroupHelper.cs [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+    protected override string MaxMemoryUsageFileName => null;
+
+    public UnidentifiedCGroup(string errorMessage)
+    {
+        _errorMessage = errorMessage;
+    }
+
+    protected override string FindCGroupPathForMemory()
+    {
+        if (_lastLog + TimeSpan.FromMinutes(10) < DateTime.UtcNow)
```
Comment (gregolsky) [inline src/Sparrow.Server/Platform/Posix/CGroupHelper.cs:328]: I think we have `SystemTime` class for UtcNow? Please consider changing DateTime occurences to that.

File: src/Sparrow.Server/Platform/Posix/CGroupHelper.cs [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+
+public static class CGroupHelper
+{
+    private static readonly Logger Logger = LoggingSource.Instance.GetLogger("Server", nameof(CGroupHelper));
+    public static readonly CGroup CGroup = GetCGroup();
+
+    private static CGroup GetCGroup()
+    {
+        const uint TMPFS_MAGIC = 0x01021994;
+        const uint CGROUP2_SUPER_MAGIC = 0x63677270;
```
Comment (gregolsky) [inline src/Sparrow.Server/Platform/Posix/CGroupHelper.cs:346]: Magic 🧙‍♂️

Comment (haludi) [reply src/Sparrow.Server/Platform/Posix/CGroupHelper.cs:346]: I just used the same name
https://github.com/omajid/dotnet-runtime/blob/392dd8b01aab6b3731fe7e5fbacdfd218c57da78/src/coreclr/src/gc/unix/cgroup.cpp#L41

File: src/Sparrow.Server/Platform/Posix/CGroupHelper.cs (hunk 3 above)
Comment (haludi) [reply src/Sparrow.Server/Platform/Posix/CGroupHelper.cs:328]: I see it belongs to `Raven.Client` and the `Sparrow.Server` doesn't have a reference to it.
Moving that will add a lot of mess to the PR and I don't see much benefit so I think it doesn't worth it.

**Comment by haludi** [comment]
> Can we add a smoke test (happy path) that would run on Linux only and try to read those values?

Done

File: src/Sparrow.Server/Platform/Posix/CGroupHelper.cs (hunk 4 above)
Comment (gregolsky) [reply src/Sparrow.Server/Platform/Posix/CGroupHelper.cs:346]: Yeah, all good :)

File: src/Sparrow.Server/Platform/Posix/CGroupHelper.cs (hunk 3 above)
Comment (gregolsky) [reply src/Sparrow.Server/Platform/Posix/CGroupHelper.cs:328]: Understood. That ok with you @ppekrol ?

**Comment by ppekrol** [comment]
The test that you added failed in GH

File: src/Sparrow.Server/Platform/Posix/CGroupHelper.cs (hunk 3 above)
Comment (ppekrol) [reply src/Sparrow.Server/Platform/Posix/CGroupHelper.cs:328]: Fine

**Comment by haludi** [comment]
cgroup v2 `memory.peak` file was added later. 
So if this value is not available we just ignore it.
https://github.com/torvalds/linux/commit/8e20d4b332660a32e842e20c34cfc3b3456bc6dc

**Comment by gregolsky** [comment]
Is this code in-sync with latest dotnet/runtime repo approach? They probably took that into account, no?

**Comment by haludi** [comment]
> Is this code in-sync with latest dotnet/runtime repo approach? They probably took that into account, no?

They didn't use this value.

**Review by gregolsky — APPROVED** [review]

**Pull request merged**